File: sp/poker/server/src/main/cpp/server/connection/message/message.cpp
#include "message.hpp"


const std::shared_ptr<header> &message::get_header() const {
    return _header;
}

void message::set_header(const std::shared_ptr<header> &header) {
    message::_header = header;
}

const std::shared_ptr<payload> &message::get_payload() const {
    return _payload;
}

void message::set_payload(const std::shared_ptr<payload> &payload) {
    message::_payload = payload;
}

message::message(std::shared_ptr<header> header_, std::shared_ptr<payload> payload_) {
    this->_header = header_;
    this->_payload = payload_;
}

std::string message::to_string() {
    return "Message: {" + _header->to_string() + "; " + _payload->to_string() + "}";
}

std::string message::construct() {
    std::string payload_str = _payload->construct();
    _header->set_length(payload_str.length());
    std::string header_str = _header->construct();

    return header_str + payload_str;
}

File: sp/poker/client/pocker-client/src/main/java/com/vpavlov/pockerclient/connection/Message.java
package com.vpavlov.pockerclient.connection;

import com.vpavlov.pockerclient.connection.header.Header;
import com.vpavlov.pockerclient.connection.payload.Payload;
import com.vpavlov.pockerclient.connection.payload.mapping.PayloadMapper;

public class Message {

    private Header header;
    private Payload payload;

    public Message(Header header, Payload payload) {
        this.header = header;
        this.payload = payload;
    }

    public Message(){}

    public void setHeader(Header header) {
        this.header = header;
    }

    public void setPayload(Payload payload) {
        this.payload = payload;
    }

    public Header getHeader() {
        return header;
    }

    public Payload getPayload() {
        return payload;
    }

    public String construct(){
        String payloadConstruct = PayloadMapper.map(this.payload);
        this.header.setLength(payloadConstruct.length());
        String headerConstruct = this.header.construct();
        return headerConstruct+payloadConstruct;
    }

    public static Message parse(String requestMsg){
        Header header = Header.extract(requestMsg);
        Payload payload = Payload.extract(requestMsg);

        return new Message(header, payload);
    }

    @Override
    public String toString() {
        return "Message{" +
                "header=" + header +
                ", payload=" + payload +
                '}';
    }
}
